pkg/webhook/webhook.go

```diff
@@ -0,0 +1,250 @@
+package webhook
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"net"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/kubernetes/test/utils"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	mutatingWebhookConfigurationName   = "kusionstack-rollout-mutating"
+	validatingWebhookConfigurationName = "kusionstack-rollout-validating"
+	webhookCertsSecretName             = "rollout-webhook-certs"
+)
+
+var webhookInitializerOnce sync.Once
+
+type RegisterHandlerFunc func(manager ctrl.Manager)
+
+// SetupWithManager sets up the webhook with a manager.
+func SetupWithManager(mgr ctrl.Manager, f RegisterHandlerFunc) (bool, error) {
+	f(mgr)
+
+	webhookInitializerOnce.Do(func() {
+		err := initialize(context.Background(), mgr)
+		if err != nil {
+			panic("unable to initialize webhook: " + err.Error())
+		}
+	})
+
+	return true, nil
+}
+
+// majority of the code is copied from kusionstack.io/operating/pkg/webhook/webhook.go
+func initialize(ctx context.Context, mgr ctrl.Manager) error {
+	logger := mgr.GetLogger().WithName("webhook")
+
+	clientset, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+
+	secret, err := ensureWebhookSecret(ctx, clientset, mgr.GetWebhookServer().Host)
+	if err != nil {
+		return err
+	}
+	logger.Info("webhook secret ensured", "namespace", secret.Namespace, "name", secret.Name)
+
+	mwhc, err := clientset.AdmissionregistrationV1().MutatingWebhookConfigurations().Get(ctx, mutatingWebhookConfigurationName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	for i := range mwhc.Webhooks {
+		if mwhc.Webhooks[i].ClientConfig.CABundle == nil {
+			mwhc.Webhooks[i].ClientConfig.CABundle = secret.Data["ca.crt"]
+		}
+	}
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
use utils.UpdateOnConflict instead of default retry

Reply:
done